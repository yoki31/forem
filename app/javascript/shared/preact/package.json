{
  "name": "preact-root-fragment",
  "version": "0.2.0",
  "main": "./preact-root-fragment.js",
  "module": "./preact-root-fragment.js",
  "type": "module"
}
