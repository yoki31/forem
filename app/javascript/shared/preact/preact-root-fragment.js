/**
 * A Preact 11+ implementation of the `replaceNode` parameter from Preact 10.
 *
 * This creates a "Persistent Fragment" (a fake DOM element) containing one or more
 * DOM nodes, which can then be passed as the `parent` argument to Preact's `render()` method.
 * Source: https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c
 */
export function createRootFragment(parent, replaceNode) {
  if (replaceNode) {
    replaceNode = Array.isArray(replaceNode) ? replaceNode : [replaceNode];
  } else {
    replaceNode = [parent];
    parent = parent.parentNode;
  }

  const s = replaceNode[replaceNode.length - 1].nextSibling;

  const rootFragment = {
    nodeType: 1,
    parentNode: parent,
    firstChild: replaceNode[0],
    childNodes: replaceNode,
    insertBefore: (c, r) => {
      parent.insertBefore(c, r || s);
      return c;
    },
    appendChild: (c) => {
      parent.insertBefore(c, s);
      return c;
    },
    removeChild: (c) => {
      parent.removeChild(c);
      return c;
    },
  };

  parent.__k = rootFragment;
  return rootFragment;
}
