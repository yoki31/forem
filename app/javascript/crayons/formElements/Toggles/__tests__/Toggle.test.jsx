import { h } from 'preact';
import { render } from '@testing-library/preact';
import { axe } from 'jest-axe';
import { Toggle } from '@crayons';

describe('<Toggle />', () => {
  it('should have no a11y violations when rendered with a label', async () => {
    const { container } = render(
      // Disabling this lint warning, as the linter doesn't know the <Toggle /> component includes the input it's looking for
      // eslint-disable-next-line jsx-a11y/label-has-associated-control
      <label>
        Example label
        <Toggle />
      </label>,
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should render default', () => {
    const { container } = render(<Toggle />);
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('should render with additional input props', () => {
    const { container } = render(
      <Toggle disabled={true} className="example-class" />,
    );
    expect(container.innerHTML).toMatchSnapshot();
  });
});
