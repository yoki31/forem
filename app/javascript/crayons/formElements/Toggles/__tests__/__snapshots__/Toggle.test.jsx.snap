// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> should render default 1`] = `"<input type=\\"checkbox\\" class=\\"c-toggle\\">"`;

exports[`<Toggle /> should render with additional input props 1`] = `"<input type=\\"checkbox\\" disabled=\\"\\" class=\\"example-class\\">"`;
